--- a/vendor/symfony/symfony/src/Symfony/Component/HttpFoundation/Request.php
+++ b/vendor/symfony/symfony/src/Symfony/Component/HttpFoundation/Request.php
@@ -2012,7 +2012,7 @@ class Request
         $pathInfo = substr($requestUri, \strlen($baseUrl));
-        if (false === $pathInfo || '' === $pathInfo) {
+        if (false === $pathInfo || '' === $pathInfo || '/' !== $pathInfo[0]) {
             // If substr() returns false then PATH_INFO is set to an empty string
-            return '/';
+            return '/'.$pathInfo;
         }
