--- a/vendor/symfony/http-foundation/Request.php
+++ b/vendor/symfony/http-foundation/Request.php
@@ -1984,9 +1984,9 @@ class Request
         }

         $pathInfo = substr($requestUri, \strlen($baseUrl));
-        if (false === $pathInfo || '' === $pathInfo) {
+        if (false === $pathInfo || '' === $pathInfo || '/' !== $pathInfo[0]) {
             // If substr() returns false then PATH_INFO is set to an empty string
-            return '/';
+            return '/'.$pathInfo;
         }

         return $pathInfo;

