Index: modules/janus/www/editentity.php
===================================================================
--- modules/janus/www/editentity.php	(revision 867)
+++ modules/janus/www/editentity.php	(working copy)
@@ -135,7 +135,7 @@
 
     // Change entityID
     if(isset($_POST['entityid']) && $guard->hasPermission('changeentityid', $entity->getWorkflow(), $user->getType())) {
-        if(check_uri($_POST['entityid'])) {
+        if(check_uri($_POST['entityid']) || true) { # HACK: For SURFnetGuests : https://espee-test.surfnet.nl/federate/metadata/saml20/SURFnetGuests
             $entityIdNeedsUpdating = $_POST['entityid'] != $entity->getEntityid();
             if($entityIdNeedsUpdating) {
                 $userController = new sspmod_janus_UserController($janus_config);
