Index: modules/janus/templates/editentity.php
===================================================================
--- modules/janus/templates/editentity.php	(revision 685)
+++ modules/janus/templates/editentity.php	(revision )
@@ -12,9 +12,14 @@
 $ssp_config = SimpleSAML_Configuration::getConfig();
 $this->data['jquery'] = array('version' => '1.6', 'core' => TRUE, 'ui' => TRUE, 'css' => TRUE);
 $this->data['head']  = '<link rel="stylesheet" type="text/css" href="/' . $this->data['baseurlpath'] . 'module.php/janus/resources/style.css" />' . "\n";
+$this->data['head'] .= '<link rel="stylesheet" type="text/css" href="/' . $this->data['baseurlpath'] . 'module.php/serviceregistry/resources/styles/serviceregistry.validate.css" />'."\n";
 $this->data['head'] .= '<script type="text/javascript" src="/' . $this->data['baseurlpath'] . 'module.php/janus/resources/scripts/swfupload.js"></script>' . "\n";
 $this->data['head'] .= '<script type="text/javascript" src="/' . $this->data['baseurlpath'] . 'module.php/janus/resources/scripts/jquery-asyncUpload-0.1.js"></script>' . "\n";
 $this->data['head'] .= '<script type="text/javascript" src="/' . $this->data['baseurlpath'] . 'module.php/janus/resources/scripts/json2-min.js"></script>'."\n";
+$this->data['head'] .= '<script type="text/javascript" src="/' . $this->data['baseurlpath'] . 'module.php/serviceregistry/resources/scripts/jquery.tmpl.min.js"></script>'."\n";
+$this->data['head'] .= '<script type="text/javascript" src="/' . $this->data['baseurlpath'] . 'module.php/serviceregistry/resources/scripts/datehelper.js"></script>'."\n";
+$this->data['head'] .= '<script type="text/javascript" src="/' . $this->data['baseurlpath'] . 'module.php/serviceregistry/resources/scripts/serviceregistry.validate.js"></script>'."\n";
+$this->data['head'] .= '<script type="text/javascript" src="/' . $this->data['baseurlpath'] . 'module.php/serviceregistry/resources/scripts/serviceregistry.validate.metadata.js"></script>'."\n";
 $this->data['head'] .= '<script type="text/javascript">
 <!--
 $(document).ready(function() {
@@ -80,7 +85,7 @@
     });
 
     // ARP edit
-    $("#arp_edit_close").click(function(){  
+    $("#arp_edit_close").click(function(){
         disablePopup();  
     });  
     $("#arp_edit_close").hover(
@@ -96,7 +101,19 @@
 });
 -->
 </script>';
+$this->data['head'] .= '
+<style>
+li, ul {
+    list-style: none;
+    margin: 0 0 0 10px;
+}
 
+ul {
+    margin: 0;
+}
+
+</style>';
+
 $this->includeAtTemplateBase('includes/header.php');
 $util = new sspmod_janus_AdminUtil();
 $wfstate = $this->data['entity_state'];
@@ -128,6 +145,7 @@
     <li><a href="#addmetadata"><?php echo $this->t('tab_import_metadata'); ?></a></li>
     <li><a href="#history"><?php echo $this->t('tab_edit_entity_history'); ?></a></li>
     <li><a href="#export"><?php echo $this->t('tab_edit_entity_export'); ?></a></li>
+    <li><a id="validate_link" href="#validate">Validate</a></li>
 </ul>
 <!-- TABS END -->
 
@@ -1111,10 +1129,18 @@
     ?>
     <table>
         <tr>
-            <td>XML:</td>
+            <td style="width: 40px">XML:</td>
             <td><textarea name="meta_xml" cols="80" rows="20"></textarea></td>
         </tr>
     </table>
+
+    <h2><?php echo $this->t('tab_edit_entity_import_json'); ?></h2>
+    <table>
+        <tr>
+            <td style="width: 40px">JSON:</td>
+            <td><textarea name="meta_json" cols="80" rows="3"></textarea></td>
+        </tr>
+    </table>
     <?php
     } else {
         echo $this->t('error_import_metadata_permission');
@@ -1132,6 +1158,149 @@
 }
 ?>
 </div>
+<!-- VALIDATE TAB -->
+<div id="validate">
+    <h2>Metadata Validation</h2>
+    <div id="MetadataValidation" class="<?php echo $this->data['entity']->getEntityid() ?>">
+        <div class="metadata-messages messages">
+        </div>
+        <script class="metadata-messages-template" type="text/x-jquery-tmpl">
+            {{each Errors}}
+            <p class="error">${$value}</p>
+            {{/each}}
+            {{each Warnings}}
+            <p class="warning">${$value}</p>
+            {{/each}}
+        </script>
+        <br />
+        <div class="entity-metadata-validation">
+        </div>
+        <script class="entity-metadata-validation-template" type="text/x-jquery-tmpl">
+            <table class="entity-metadata-table">
+                <thead>
+                    <tr>
+                        <th>Entry</th>
+                        <th>Value</th>
+                    </tr>
+                </thead>
+                <tbody>
+                    {{each Validations}}
+                    <tr>
+                        <td>${name}</td>
+                        <td class="${status}">
+                            {{if message.length == 1}}
+                                {{each message}} <span class="message">${$value}</span> {{/each}}
+                            {{else}}
+                                <ul class="entity-metadata">
+                                    {{each message}}
+                                        <li>${$value}</li>
+                                    {{/each}}
+                                </ul>
+                            {{/if}}</td>
+                    </tr>
+                    {{/each}}
+                </tbody>
+            </table>
+        </script>
+    </div>
+    <h2>Certificate & Endpoints validation</h2>
+    <ul>
+        <li class="entity-type">
+            <ul>
+                <li class="entity">
+                    <div class="entity-messages messages">
+                    </div>
+
+                    <script class="messages-template" type="text/x-jquery-tmpl">
+                        {{each Errors}}
+                        <p class="error">${$value}</p>
+                        {{/each}}
+                        {{each Warnings}}
+                        <p class="warning">${$value}</p>
+                        {{/each}}
+                    </script>
+
+                    <table class="entity-information">
+                        <tr>
+                            <th>Entity ID</th>
+                            <td>
+                                <a href="<?php echo $this->data['entity']->getEntityid() ?>" class="entity-id">
+                                    <?php echo $this->data['entity']->getEntityid() ?>
+                                </a>
+                            </td>
+                        </tr>
+                        <tr>
+                            <th>Metadata URL</th>
+                            <td>
+                                <a href="<?php echo $this->data['entity']->getMetadataURL() ?>">
+                                    <?php echo $this->data['entity']->getMetadataURL() ?>
+                                </a>
+                            </td>
+                        </tr>
+                    </table>
+
+                    <br />
+
+                    <p class="header-25">
+                        Signing Certificate
+                    </p>
+
+                    <div class="entity-certificate-representation">
+                    </div>
+
+                    <div class="entity-certificate-information">
+                        <img class="loading-image" alt='Loading...' src="/simplesaml/module.php/serviceregistry/resources/images/icons/spinner.gif" />
+                    </div>
+
+                    <script class="entity-certificate-information-template" type="text/x-jquery-tmpl">
+                        <table>
+                            <tr>
+                                <th>Subject:</th>
+                                <td>${Subject}</td>
+                            </tr>
+                            <tr>
+                                <th>Starts / started:</th>
+                                <td>${Starts_natural} (${Starts_relative})</td>
+                            </tr>
+                            <tr>
+                                <th>Ends / ended:</th>
+                                <td>${Ends_natural} (${Ends_relative})</td>
+                            </tr>
+                        </table>
+                    </script>
+
+                    <br />
+
+                    <p class="header-25">
+                        Endpoints
+                    </p>
+                    <img class="loading-image" alt='Loading...' src="/simplesaml/module.php/serviceregistry/resources/images/icons/spinner.gif" />
+                    <ul class="entity-endpoints">
+                    </ul>
+
+                    <script class="entity-endpoint-template" type="text/x-jquery-tmpl">
+                        <li>
+                            <h3>
+                                <img style="display: inline;" height="24px" width="24px" src="/simplesaml/module.php/serviceregistry/resources/images/icons/endpoint.png" alt="" />
+                                ${Name}
+                            </h3>
+                            <a href="${Url}">${Url}</a>
+
+                            <div class="entity-endpoint-messages messages">
+                            </div>
+
+                            <div class="entity-endpoint-certificate-representation">
+                            </div>
+
+                            <div class="entity-endpoint-certificate-information">
+                            </div>
+                        </li>
+                    </script>
+                </li>
+            </ul>
+        </li>
+    </ul>
+</div>
 <hr />
 <?php echo $this->t('tab_edit_entity_revision_note'); ?>: <input type="text" name="revisionnote" class="revision_note" />
 <input type="submit" name="formsubmit" id="master_submit" value="<?php echo $this->t('tab_edit_entity_save'); ?>" class="save_button"/>
