Index: modules/janus/lib/MetaExport.php
===================================================================
--- modules/janus/lib/MetaExport.php	(revision 784)
+++ modules/janus/lib/MetaExport.php	(working copy)
@@ -52,37 +52,73 @@
 
         $metadata_raw = $econtroller->getMetadata();
 
-        $metadata_required = $janus_config->getArray('metadatafields.' . $entity->getType());
+        $metadata_info = $janus_config->getArray('metadatafields.' . $entity->getType());
 
         $required = array();
-        foreach($metadata_required AS $k => $v) {
+        foreach($metadata_info AS $k => $v) {
             if(array_key_exists('required', $v) && $v['required'] === true) {
                 $required[] = $k;
             }
+            // Expand fields like:
+            // AssertionConsumerService:#:Binding
+            // with:
+            // supported = array(0,1)
+            // to:
+            // AssertionConsumerService:0:Binding
+            // AssertionConsumerService:1:Binding
+            if (!empty($v['supported'])) {
+                foreach ($v['supported'] as $supportedValue) {
+                    $expandedKey = str_replace('#', $supportedValue, $k);
+                    $expandedValue = $v;
+                    unset($expandedValue['supported']);
+                    $metadata_info[$expandedKey] = $v;
+                }
+            }
         }
-
+        
         $metadata = array();
         foreach($metadata_raw AS $k => $v) {
-            if (!isset($metadata_required[$v->getKey()])) {
+            if (!isset($metadata_info[$v->getKey()])) {
                 continue;
             }
-            if ($v->getValue() == '') {
+            if ((string)$v->getValue() === '') {
                 continue;
             }
 
             $default_allow = false;
-            if(isset($metadata_required[$v->getKey()]['default_allow']) && is_bool($metadata_required[$v->getKey()]['default_allow'])) {
-                $default_allow = $metadata_required[$v->getKey()]['default_allow'];
+            if(isset($metadata_info[$v->getKey()]['default_allow']) && is_bool($metadata_info[$v->getKey()]['default_allow'])) {
+                $default_allow = $metadata_info[$v->getKey()]['default_allow'];
             }
 
-            if (!$default_allow && (isset($metadata_required[$v->getKey()]['default']) && ($v->getValue() == $metadata_required[$v->getKey()]['default']))) {
+            if (!$default_allow && (isset($metadata_info[$v->getKey()]['default']) && ((string)$v->getValue() === (string)$metadata_info[$v->getKey()]['default']))) {
                 continue;
             }
 
             $metadata[] = $v->getKey();
         }
 
-        $missing_required = array_diff($required, $metadata);
+        // Check if we are missing required metadata fields,
+        $missing_required = $required;
+        foreach ($missing_required as $required_index => $required_key) {
+            // Expand fields with # syntax
+            $required_key_expanded = array();
+            if (strstr($required_key, '#') && !empty($metadata_info[$required_key]['supported'])) {
+                foreach ($metadata_info[$required_key]['supported'] as $supported_value) {
+                    $required_key_expanded[] = str_replace('#', $supported_value, $required_key);
+                }
+            }
+
+            foreach ($metadata as $metadata_key) {
+                if ($metadata_key === $required_key) {
+                    unset ($missing_required[$required_index]);
+                    continue 2; // Required field found, continue with the next required field
+                }
+                else if (in_array($metadata_key, $required_key_expanded)) {
+                    unset($missing_required[$required_index]);
+                    continue 2; // Required field found, continue with the next required field
+                }
+            }
+        }
         
         $entityid = $entity->getEntityid();
         
