Index: www/dashboard.php
===================================================================
--- www/dashboard.php	(revision 498)
+++ www/dashboard.php	(working copy)
@@ -136,6 +136,7 @@
 if(isset($_POST['usersubmit'])) {
     $user->setData($_POST['userdata']);
     $user->setEmail($_POST['user_email']);
+    $user->setSecret($_POST['user_secret']);
     $user->save();
     $pm->post(
         'Userinfo update',
Index: dictionaries/janus.definition.json
===================================================================
--- dictionaries/janus.definition.json	(revision 530)
+++ dictionaries/janus.definition.json	(working copy)
@@ -83,6 +83,9 @@
 	"tab_user_data_email": {
 		"en": "E-mail"
 	},
+	"tab_user_data_secret": {
+	    "en": "Secret (for REST access)"
+	},
 	"tab_user_data_username": {
 		"en": "User ID"
 	},
Index: dictionaries/janus.translation.json
===================================================================
--- dictionaries/janus.translation.json	(revision 499)
+++ dictionaries/janus.translation.json	(working copy)
@@ -111,6 +111,10 @@
 		"da": "E-mail",
 		"es": "E-mail"
 	},
+	"tab_user_data_secret": {
+	    "da": "Secret (for REST access)",
+	    "es": "Secret (for REST access)"
+	},
 	"tab_user_data_username": {
 		"da": "Bruger ID",
 		"es": "Nombre de usuario"
Index: templates/dashboard.php
===================================================================
--- templates/dashboard.php	(revision 498)
+++ templates/dashboard.php	(working copy)
@@ -637,6 +637,7 @@
         <h2><?php echo $this->t('tab_user_data_subheader');  ?></h2>
         <p><?php echo $this->t('tab_user_data_username');  ?>: <?php echo $this->data['user']->getUserid(); ?></p>
         <p><?php echo $this->t('tab_user_data_email');  ?>: <input type="text" name="user_email" value="<?php echo $this->data['user']->getEmail(); ?>" /></p>
+        <p><?php echo $this->t('tab_user_data_secret'); ?>: <input type="text" name="user_secret" value="<?php echo $this->data['user']->getSecret(); ?>" size="50"/></p>
         <p><?php echo $this->t('tab_user_data_otherinfo');  ?>:</p>
         <textarea name="userdata" cols="100" rows="10"><?php echo $this->data['user']->getData(); ?></textarea>
         <input type="submit" name="usersubmit" value="<?php echo $this->t('tab_edit_entity_save'); ?>" />
