Index: services/rest/index.php
===================================================================
--- services/rest/index.php	(revision 484)
+++ services/rest/index.php	(working copy)
@@ -2,9 +2,12 @@
 try {
     $request = sspmod_janus_REST_Utils::processRequest($_GET);
 
-    $result = sspmod_janus_REST_Utils::callMethod($request);
-    
-    sspmod_janus_REST_Utils::sendResponse($result['status'], $result['data'], 'application/json');
+    if (is_object($request)) {
+        $result = sspmod_janus_REST_Utils::callMethod($request);
+        sspmod_janus_REST_Utils::sendResponse($result['status'], $result['data'], 'application/json');
+    } else {
+        sspmod_janus_REST_Utils::sendResponse(500);
+    }
 } catch(Exception $e) {
     sspmod_janus_REST_Utils::sendResponse(500);
 }
