Index: AdminUtil.php
===================================================================
--- AdminUtil.php       (revision 484)
+++ AdminUtil.php       (working copy)
@@ -113,7 +113,7 @@
             'SELECT `eid`, `entityid`, MAX(`revisionid`) AS `revisionid`,
                 `created`
             FROM `'. self::$prefix .'entity` WHERE ' . implode(' AND ', $sql) . '
-            GROUP BY `entityid`;', 
+            GROUP BY `eid`;', 
             $params
         );

