Index: www/editentity.php
===================================================================
--- www/editentity.php	(revision 484)
+++ www/editentity.php	(working copy)
@@ -35,6 +35,10 @@
 $user->setUserid($userid);
 $user->load(sspmod_janus_User::USERID_LOAD);
 
+// Get Admin util which we use to retrieve entities
+$autil = new sspmod_janus_AdminUtil();
+
+
 // Function to fix up PHP's messing up POST input containing dots, etc.
 function getRealPOST() {
     $vars = array();
@@ -329,25 +333,42 @@
 
 // Get remote entities
 if($entity->getType() == 'saml20-sp') {
-    $remote_entities = $metadata->getList('saml20-idp-remote');
-    $remote_entities = array_merge($metadata->getList('shib13-idp-remote'), $remote_entities);
+ 
+    $loaded_entities = array_merge($autil->getEntitiesByStateType(null, 'saml20-idp'),
+                                   $autil->getEntitiesByStateType(null, 'shib13-idp'));
     $et->data['metadata_fields'] = $janus_config->getValue('metadatafields.saml20-sp');
+    
+    
 } else if($entity->getType() == 'saml20-idp') {
-    $remote_entities = $metadata->getList('saml20-sp-remote');
-    $remote_entities = array_merge($metadata->getList('shib13-sp-remote'), $remote_entities);
+    $loaded_entities = array_merge($autil->getEntitiesByStateType(null, 'saml20-sp'),
+                                   $autil->getEntitiesByStateType(null, 'shib13-sp'));
     $et->data['metadata_fields'] = $janus_config->getValue('metadatafields.saml20-idp');
 } else if($entity->getType() == 'shib13-sp') {
-    $remote_entities = $metadata->getList('saml20-idp-remote');
-    $remote_entities = array_merge($metadata->getList('shib13-idp-remote'), $remote_entities);
+
+    $loaded_entities = array_merge($autil->getEntitiesByStateType(null, 'saml20-idp'),
+                                   $autil->getEntitiesByStateType(null, 'shib13-idp'));
     $et->data['metadata_fields'] = $janus_config->getValue('metadatafields.saml20-sp');
 } else if($entity->getType() == 'shib13-idp') {
-    $remote_entities = $metadata->getList('saml20-sp-remote');
-    $remote_entities = array_merge($metadata->getList('shib13-sp-remote'), $remote_entities);
+    $loaded_entities = array_merge($autil->getEntitiesByStateType(null, 'saml20-sp'),
+                                   $autil->getEntitiesByStateType(null, 'shib13-sp'));    
     $et->data['metadata_fields'] = $janus_config->getValue('metadatafields.saml20-idp');
 }
 
+$remote_entities = array();
+
 // Only parse name and description in current language
-foreach($remote_entities AS $key => $value) {
+foreach($loaded_entities AS $entityRow) {
+    
+    $instance = new sspmod_janus_Entity($janus_config);
+    $instance->setEid($entityRow["eid"]);
+    $instance->setRevisionid($entityRow["revisionid"]);
+    $instance->load();
+    
+    $value = array("name"=>$instance->getPrettyName(),
+                   "description"=>$instance->getEntityId(),
+                   );
+    $key = $instance->getEntityId();
+        
     unset($value2);
     if(isset($value['name'])) {
         if(is_array($value['name'])) {
