Index: modules/janus/templates/dashboard.php
===================================================================
--- modules/janus/templates/dashboard.php	(revision 784)
+++ modules/janus/templates/dashboard.php	(working copy)
@@ -603,6 +603,9 @@
     <?php
     echo '<h2>'.$this->t('tab_entities_federation_entity_subheader').'</h2>';
     echo '<a href="exportentities.php">'.$this->t('tab_entities_federation_exporting').'</a>';
+    echo '<br /><a href="' . SimpleSAML_Module::getModuleURL('serviceregistry/show-entities-validation.php') . '">';
+    echo $this->t('tab_entities_federation_status');
+    echo '</a>';
     if($this->data['uiguard']->hasPermission('experimental', null, $this->data['user']->getType(), TRUE)) {
         echo '<br /><a href="metalisting.php">'.$this->t('tab_entities_federation_status').'</a><br />';
     }
