Index: UserController.php
===================================================================
--- UserController.php	(revision 484)
+++ UserController.php	(working copy)
@@ -373,10 +373,10 @@
 
         $st = $this->execute(
             'SELECT DISTINCT eid 
-            FROM '. self::$prefix ."metadata
+            FROM '. self::$prefix ."metadata jm
             WHERE `key` = ?
             AND ((value=?) OR (? REGEXP CONCAT('^',value,'\$')))
-            AND revisionid = (SELECT MAX(revisionid) FROM ".self::$prefix."metadata md WHERE md.eid = eid);",
+            AND revisionid = (SELECT MAX(revisionid) FROM ".self::$prefix."metadata WHERE eid = jm.eid);",
                 array($key, $value, $value)
             );
 
